### DISCUSSION TITLE
Setting RENOVATE_HOST_RULES in Gitlab and keeping it masked?

### DISCUSSION THREAD

**Discussion by kastl-ars** [renovatebot/renovate#29898] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Gitlab Pipeline, using https://gitlab.com/renovate-bot/renovate-runner/

### Please tell us more about your question or problem

I am using renovate inside a Gitlab Pipeline by including things from https://gitlab.com/renovate-bot/renovate-runner/.

It works, but I need to set a hostRule. From [what I found](https://docs.renovatebot.com/getting-started/private-packages/) it seems that I can do that by setting a CI/CD Pipeline variable `RENOVATE_HOST_RULES` to a json string like this:

```
{
  "hostRules": [
    {
      "matchHost": "https://registry.company.com/pypi-simple/",
      "username": "engineering",
      "password": "abc123"
    }
  ]
}
```

But Gitlab warns me that it cannot mask the variable due to characters it found (whitespace, braces, etc.). So I might end up publishing the token.

Hardcoding the token in the repository is of course a no-go.

Having a step before running renovate that somehow creates the `config.json` by `cat`-ing things into a file, while using environment variables from the Gitlab Pipeline, might be possible, but also sounds like a hack.

Is there a way to only set the Token in a Gitlab variable and tell renovate to use **that** token for **that** host?

Did anyone solve this in a nice way? Or am I being paranoid here?

Thanks in advance,
Johannes

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by RahulGautamSingh** [comment, marked as answer]
Create a config.js file in the root of the repo (runner) -> add the `hostRules` in it -> pass the token via env. 
You can read from there how to do it correctly and how it works:
https://gitlab.com/renovate-bot/renovate-runner#other-config-options -> https://github.com/renovatebot/docker-renovate/blob/HEAD/docs/gitlab.md#renovate-config

**Reply by kastl-ars** [reply]
Thanks, I'll take a look.

**Reply by kastl-ars** [reply]
Hmmm, do I need to add a complete config.js or is it enough to just set the hostRules (to not overwrite the settings used automatically by the pipeline)?

**Comment by rarkins** [comment]
Alternatively consider the use of `secrets` and then your `hostRules` themselves don't contain any direct secrets

**Reply by kastl-ars** [reply]
Thanks for the reply! Could you elaborate on this a little bit?

You mean this documentation: https://docs.renovatebot.com/self-hosted-configuration/#secrets

How would I use that with Gitlab as there is no config.js being used? Or at least none defined by me (yet)? How would I create it and populate it with secrets from Gitlab?

**Reply by rarkins** [reply]
I mean this configuration option: https://docs.renovatebot.com/self-hosted-configuration/#secrets

**Reply by kastl-ars** [reply]
Sorry, but I still fail to see how I would use that? If I set a `RENOVATE_SECRETS` variable to contain a module.exports block, I will end up with the same warning, that the variable might not be masked (due to the variable containing funny characters as it is json...).

**Reply by rarkins** [reply]
then use config.js if you prefer. You can combine file and env config

**Reply by kastl-ars** [reply]
Sorry, now I am confused. Maybe I am lacking coffee, but for GitlabCI pipelines I have no way to use a config.js containing secrets. I can not commit it to the repository, right?

I'll try to set the hostRules in the config.js with using just the token from a Gitlab secret, this way the token should be properly masked...

**Reply by rarkins** [reply]
You should use a centralized pipeline to run against many repos. And then you commit the config.js to that central repo

**Reply by kastl-ars** [reply]
Actually this is what I am trying right now, but still, the config.js should not contain the secrets.

This seems to be working (but fails due to other timeout issues unreleated to the hostRules):

```
Object.assign(process.env);

module.exports = {
  "hostRules": [
    {
      "matchHost": "artifactory.example.com",
      "username": process.env.ARTIFACTORY_USER,
      "password": process.env.ARTIFACTORY_TOKEN
    }
  ]
}
```
